Unable to format due to bug

This template is a guideline, not a strict requirement.
- **Version**: 2.6.4
- **Integration**: CLI
- **Configuration**:
```
version=2.6.4
newlines.neverInResultType = true
```

## Steps

Given code like this:
```scala
object a {
implicit protected val td: TildeArrow[RequestContext, Future[RouteResult]] {
    type Out = RouteTestResult
  } = TildeArrow.injectIntoRoute

  }
```

When I run scalafmt like this:

```bash
scalafmt
```

## Problem

Scalafmt formats code like this:
```
org.scalafmt.cli.FailedToFormat: /private/tmp/a.scala
Caused by: org.scalafmt.Error$PreciseIncomplete: /private/tmp/a.scala:2: error: Unable to format file due to bug in scalafmt
implicit protected val td: TildeArrow[RequestContext, Future[RouteResult]] {
                                                                           ^
```

## Workaround

Disable `neverInResultType`.

## Notes

- [The only place](https://github.com/scalameta/scalafmt/blob/af2205e38716474c39a9a8933547f2cb8772e030/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala#L999-L1003) of config usage
- Also `neverInResultType` config is not documented yet.

--- a/docs/configuration.md
+++ b/docs/configuration.md
@@ -1359,6 +1359,30 @@ class Engine[TD, EI, PD, Q, P, A](
       Class[_ <: BaseDataSource[TD, EI, Q, A]]]) {}
 ```
 
+### `newlines.avoidInResultType`
+
+If true, newlines in definition result type will only be used if formatting
+without them is impossible. This parameter was added in 2.7.0, replacing
+`neverInResultType`.
+
+```scala mdoc:defaults
+newlines.avoidInResultType
+```
+
+```scala mdoc:scalafmt
+maxColumn = 40
+newlines.avoidInResultType = true
+newlines.neverBeforeJsNative = true
+---
+// no newlines in result type
+def permissionState(a: A = js.native): js.Promise[PushPermissionState] = js.native
+// no newlines in result type
+val permissionState: js.Promise[PushPermissionState] = js.native
+// can't format without newlines
+implicit protected val td: TildeArrow {
+  type Out = RouteTestResult } = TildeArrow.injectIntoRoute
+```
+
 ## Rewrite Rules
 
 To enable a rewrite rule, add it to the config like this

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/config/Newlines.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/config/Newlines.scala
@@ -147,7 +147,8 @@ import metaconfig.generic.Surface
   */
 case class Newlines(
     source: SourceHints = Newlines.classic,
-    neverInResultType: Boolean = false,
+    @annotation.ExtraName("neverInResultType")
+    avoidInResultType: Boolean = false,
     neverBeforeJsNative: Boolean = false,
     sometimesBeforeColonInMethodReturnType: Boolean = true,
     penalizeSingleSelectMultiArgList: Boolean = true,

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala
@@ -266,7 +266,7 @@ object ScalafmtConfig {
     continuationIndent = ContinuationIndent(4, 4),
     importSelectors = ImportSelectors.binPack,
     newlines = default.newlines.copy(
-      neverInResultType = true,
+      avoidInResultType = true,
       neverBeforeJsNative = true,
       sometimesBeforeColonInMethodReturnType = false
     ),

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala
@@ -997,9 +997,13 @@ class Router(formatOps: FormatOps) {
             .withPolicy(penalizeNewlines)
         )
       case FormatToken(T.Colon(), _, _)
-          if style.newlines.neverInResultType &&
+          if style.newlines.avoidInResultType &&
             defDefReturnType(leftOwner).isDefined =>
-        val expire = lastToken(defDefReturnType(leftOwner).get)
+        val expire = defDefReturnType(leftOwner).get match {
+          case Type.Refine(_, headStat :: _) =>
+            tokens(headStat.tokens.head, -1).left
+          case t => lastToken(t)
+        }
         Seq(Split(Space, 0).withPolicy(SingleLineBlock(expire, okSLC = true)))
 
       case FormatToken(T.LeftParen(), T.LeftBrace(), between) =>

--- a/scalafmt-tests/src/test/resources/unit/DefDef.stat
+++ b/scalafmt-tests/src/test/resources/unit/DefDef.stat
@@ -156,8 +156,8 @@ val a = new a[b, c](/*c1*/d => e)(/*c2*/f => g, h => i)(/*c3*/j => k)
 val a = new a[b, c]( /*c1*/ d => e)(
     /*c2*/ f => g,
     h => i)( /*c3*/ j => k)
-<<< #2179 neverInResultType=true
-newlines.neverInResultType = true
+<<< #2179 avoidInResultType=true
+newlines.avoidInResultType = true
 ===
 object a {
   implicit protected val td: TildeArrow[RequestContext, Future[RouteResult]] {
@@ -165,11 +165,14 @@ object a {
   } = TildeArrow.injectIntoRoute
 }
 >>>
-BestFirstSearch:330 Failed to format
-UNABLE TO FORMAT,
-tok={∙type[89:98]
-<<< #2179 neverInResultType=false
-newlines.neverInResultType = false
+object a {
+  implicit protected val td
+      : TildeArrow[RequestContext, Future[RouteResult]] {
+        type Out = RouteTestResult
+      } = TildeArrow.injectIntoRoute
+}
+<<< #2179 avoidInResultType=false
+newlines.avoidInResultType = false
 ===
 object a {
   implicit protected val td: TildeArrow[RequestContext, Future[RouteResult]] {

--- a/scalafmt-tests/src/test/resources/unit/DefDef.stat
+++ b/scalafmt-tests/src/test/resources/unit/DefDef.stat
@@ -156,3 +156,32 @@ val a = new a[b, c](/*c1*/d => e)(/*c2*/f => g, h => i)(/*c3*/j => k)
 val a = new a[b, c]( /*c1*/ d => e)(
     /*c2*/ f => g,
     h => i)( /*c3*/ j => k)
+<<< #2179 neverInResultType=true
+newlines.neverInResultType = true
+===
+object a {
+  implicit protected val td: TildeArrow[RequestContext, Future[RouteResult]] {
+    type Out = RouteTestResult
+  } = TildeArrow.injectIntoRoute
+}
+>>>
+BestFirstSearch:330 Failed to format
+UNABLE TO FORMAT,
+tok={∙type[89:98]
+<<< #2179 neverInResultType=false
+newlines.neverInResultType = false
+===
+object a {
+  implicit protected val td: TildeArrow[RequestContext, Future[RouteResult]] {
+    type Out = RouteTestResult
+  } = TildeArrow.injectIntoRoute
+}
+>>>
+object a {
+  implicit protected val td
+      : TildeArrow[
+          RequestContext,
+          Future[RouteResult]] {
+        type Out = RouteTestResult
+      } = TildeArrow.injectIntoRoute
+}

